prov/efa/src/efa_mr.c

```diff
@@ -584,6 +584,62 @@ static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr
 				mr_attr->mr_iov->iov_len, access);
 	}
 }
+#elif HAVE_NEURON
+/**
+ * @brief Register a memory buffer with rdma-core api.
+ *
+ * @param efa_mr the ptr to the efa_mr object
+ * @param mr_attr the ptr to the fi_mr_attr object
+ * @param access the desired memory protection attributes
+ * @return struct ibv_mr* the ptr to the registered MR
+ */
+static struct ibv_mr *efa_mr_reg_ibv_mr(struct efa_mr *efa_mr, struct fi_mr_attr *mr_attr, int access)
+{
+	int dmabuf_fd;
+	struct ibv_mr *ibv_mr;
+	if (efa_mr_is_neuron(efa_mr)) {
+		int ret;
+		ret = neuron_get_dmabuf_fd(
+				(uint64_t) mr_attr->mr_iov->iov_base,
+				(uint64_t) mr_attr->mr_iov->iov_len,
+				&dmabuf_fd);
+		if (ret == FI_SUCCESS) {
+			ibv_mr = ibv_reg_dmabuf_mr(
+						efa_mr->domain->ibv_pd, 0,
+						mr_attr->mr_iov->iov_len,
+						(uint64_t)mr_attr->mr_iov->iov_base,
+						dmabuf_fd, access);
+			if (!ibv_mr && (errno == EOPNOTSUPP)) {
+				EFA_INFO(FI_LOG_MR,
+					"Registering dma-buf MR is not supported by the RDMA driver, "
+					"Falling back to ibv_reg_mr\n");
+				ret = close(dmabuf_fd);
+				if (ret) {
+					EFA_WARN(FI_LOG_MR,
+						"Failed to close dmabuf fd of Neuron device buffer: %s\n",
+						fi_strerror(-ret));
+					return NULL;
+				}
+				goto ibv_reg_mr;
+			}
+		} else {
+			EFA_INFO(FI_LOG_MR,
+				"Unable to get dmabuf fd for Neuron device buffer, "
+				"Fall back to ibv_reg_mr\n");
+			goto ibv_reg_mr;
+		}
+	} else {
+		goto ibv_reg_mr;
+	}
+
+	return ibv_mr;
+
+ibv_reg_mr:
+	return ibv_reg_mr(
+		efa_mr->domain->ibv_pd,
+		(void *)mr_attr->mr_iov->iov_base,
+		mr_attr->mr_iov->iov_len, access);
```

Comment:
May I suggest refactoring this block to avoid `goto`?

For example:
```c
if (!efa_mr_is_neuron(efa_mr))
    return ibv_reg_mr(...)

ret = neuron_get_dmabuf_fd(
				(uint64_t) mr_attr->mr_iov->iov_base,
				(uint64_t) mr_attr->mr_iov->iov_len,
				&dmabuf_fd);
if (ret != FI_SUCCESS) {
    // Actually I'm not sure if this should happen... why do we not error out instead of fallback to ibv_reg_mr?
}

ibv_mr = ibv_reg_dmabuf_mr(
						efa_mr->domain->ibv_pd, 0,
						mr_attr->mr_iov->iov_len,
						(uint64_t)mr_attr->mr_iov->iov_base,
						dmabuf_fd, access);
// Happy path

return ibv_mr;
```

Reply:
I will refactor the code. The fallback is to cover the case where Neuron driver doesn't support dmabuf feature (`neuron_get_dmabuf_fd` fails when the Neuron driver doesn't support dmabuf). This is for backward compatibility.

Reply:
Added a new commit with the refactored code. Please have a look.